### PR TITLE
Introduce Preinstaller in installation lib

### PR DESCRIPTION
**Pull request by mjakobczyk**
**Description**

Changes proposed in this pull request:

- rewrite logic from cluster-essentials crd-install-job for installation lib
- cover installation of CRDs
- cover creation of Namespaces

**Related issue(s)**

Implements #177

### CODE REVIEW

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
+package preinstaller
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/preinstaller.go:1]: It's good to describe the package here unless you want a separate go.doc file. Take a look at other packages for examples.

File: parallel-install/pkg/preinstaller/output.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package preinstaller
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/output.go:1]: These types are used only by things in "preinstaller.go". I think it makes sense to move them there.

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		applier:       applier,
+		cfg:           cfg,
+		dynamicClient: dynamicClient,
+		retryOptions:  retryOptions,
+	}
+}
+
+// InstallCRDs on a k8s cluster.
+// Returns Output containing results of installation.
+func (i *PreInstaller) InstallCRDs() (Output, error) {
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/preinstaller.go:32]: Do we consider "UninstallCRDs" operation? I think uninstall should be supported by the library.

File: parallel-install/pkg/preinstaller/preinstaller.go (hunk 1 above)
Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/preinstaller.go:1]: Will do!

File: parallel-install/pkg/preinstaller/preinstaller.go (hunk 3 above)
Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/preinstaller.go:32]: So far only logic of resources installation (such as crds or namespaces) has been considered in scope of this task, as this is what was previously done by installer job. I haven't seen anything connected with deinstallation. Tbh your suggestion is a matter of broader Kyma "uninstallation" topic. Even though it was not mentioned in the acceptance criteria I will verify it with PO.

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+func (i *PreInstaller) apply(resourceType resourceType) (o Output, err error) {
+	installationResourcePath := i.cfg.InstallationResourcePath
+	path := fmt.Sprintf("%s/%s", installationResourcePath, resourceType.name)
+
+	rawComponentsDir, err := ioutil.ReadDir(path)
+	if err != nil {
+		return o, err
+	}
+
+	components := findOnlyDirectoriesAmong(rawComponentsDir)
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/preinstaller.go:69]: What for is a "component" here? I am not sure if we need any "components" in preinstall phase.

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	if components == nil || len(components) == 0 {
+		i.cfg.Log.Warn("There were no components detected for installation. Skipping.")
+		return o, nil
+	}
+
+	for _, component := range components {
+		componentName := component.Name()
+		i.cfg.Log.Infof("Processing component: %s", componentName)
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/preinstaller.go:77]: I think that context is missing here: we don't know which resourceType is being processed. So, if we have multiple resource types, we'll have few identical log entries: "Processing component: Foo", one for every resource type.

File: parallel-install/pkg/preinstaller/preinstaller.go (hunk 4 above)
Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/preinstaller.go:69]: In current approach you can find preinstaller resources in Kyma repository. They are as it follows:
```
- installation
    - resources
        - crds
            - component1
                - file 1
                - file 2
                - ...
                - file n
            - component1
                - file 1
                - file 2
                - ...
                - file n
        - namespaces
            // the same composition as for crds
```
I have introduced a mechanism for creating resources of a specific type. They should always be under appropriate subdirectory placed in `installation/resources`. However, inside these directories resources should be divided by component (for better logical division, ease of component-flagging mechanism that could be introduced etc).
So, to sum up - in this context `component` means either internal Kyma component or external Kyma dependency (which should be still so-called components).

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+	output, err := i.apply(resource)
+	if err != nil {
+		return Output{}, err
+	}
+
+	return output, nil
+}
+
+func (i *PreInstaller) apply(resourceType resourceType) (o Output, err error) {
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/preinstaller.go:60]: I think this function does a bit too much. It is responsible for **finding** resources and for **applying** them.
How about splitting this functionality?
I'd suggest to have two functions:
findResources that does the finding/directory reads and returns a list of objects with fields like {resourceType, component, fileName, path} (only necessary data)
applyResources that is processing the list produced by findResources

I would also split this nested for loop into two private functions each having a single for loop. But I will not insist on that, if the code is simple, a one-level nesting is still OK.

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+				o.notInstalled = append(o.notInstalled, file)
+				return o, err // TODO: fail-fast or continue?
+			}
+
+			applied, err := i.applier.Apply(string(resourceData))
+			if err != nil {
+				o.notInstalled = append(o.notInstalled, file)
+				i.cfg.Log.Warnf("Error occurred when processing file %s : %s", resourceName, err.Error())
+			} else {
+				if applied {
```
[the remark this answers is not in the document]
Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/preinstaller.go:109]: In fact it should be considered for adding it to uninstalled ones. Will fix it!

File: parallel-install/pkg/preinstaller/resourceapplier.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+// NewGenericResourceApplier returns a new instance of GenericResourceApplier.
+func NewGenericResourceApplier(log logger.Interface, resourceManager ResourceManager) *GenericResourceApplier {
+	return &GenericResourceApplier{
+		log:             log,
+		decoder:         initializeDecoder(),
+		resourceManager: resourceManager,
+	}
+}
+
+func (c *GenericResourceApplier) Apply(manifest string) (bool, error) {
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/resourceapplier.go:43]: I am missing a clear contract for this method.
It returns "false, nil" in four cases. Three of these are related to parsing/decoding input files. The fourth one is when the resource already exists... This is very dangerous for upgrades. We may silently proceed with an upgrade even if one CRD was not updated to the latest version because it already exists in the cluster.
I would say we should either:
- decide it's an error when resource already exist
- return a specific, unique value that allows the caller to recognize the "resource already exists" condition and leave the decision "what to do next" to the caller.

Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/resourceapplier.go]: Agree. Contract has been changed. Now it will return error only, so fail-fast approach will be taken.

File: parallel-install/pkg/preinstaller/preinstaller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/preinstaller.go]: Contract has been changed, so this refactor is no longer applicable.

File: parallel-install/pkg/preinstaller/preinstaller.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+
+type resourceInfoInput struct {
+	dirSuffix                string
+	resourceType             string
+	installationResourcePath string
+}
+
+type resourceInfoResult struct {
+	component    string
+	name         string
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/preinstaller.go:41]: I understand that this holds the file name? I'd rename it to `fileName` then. Just the "name" requires looking at the code to realize "name of what" is stored there.

File: parallel-install/pkg/preinstaller/resourcemanager_test.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+	retryOptions := getTestingRetryOptions(cfg)
+
+	t.Run("should update resource", func(t *testing.T) {
+		// given
+		manager := NewDefaultResourceManager(dynamicClient, retryOptions)
+		resource := unstructured.Unstructured{}
+		resourceSchema := schema.GroupVersionResource{}
+
+		// when
+		err := manager.UpdateRefreshableResource(&resource, resourceSchema)
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/resourcemanager_test.go:96]: What exactly happens here? The resource is empty and has an empty schema... I am not an expert on dynamic client but I would expect some test data here like `namespace` for example.

File: parallel-install/pkg/preinstaller/resourcemanager_test.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+		}}
+		resourceSchema := schema.GroupVersionResource{}
+		err := manager.CreateResource(&testObj, resourceSchema)
+		assert.NoError(t, err)
+
+		// when
+		obj, err := manager.GetResource(resourceName, resourceSchema)
+
+		// then
+		expectedError := "not found"
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/resourcemanager_test.go:74]: The test case name suggest we should expect success, not an error.

File: parallel-install/pkg/preinstaller/resourcemanager_test.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+	retryOptions := getTestingRetryOptions(cfg)
+
+	t.Run("should create resource", func(t *testing.T) {
+		// given
+		manager := NewDefaultResourceManager(dynamicClient, retryOptions)
+		resource := unstructured.Unstructured{}
+		resourceSchema := schema.GroupVersionResource{}
+
+		// when
+		err := manager.CreateResource(&resource, resourceSchema)
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/resourcemanager_test.go:28]: What exactly happens here? The resource is empty and has an empty schema... I am not an expert on dynamic client but I would expect some test data here like a `namespace` for example.

File: parallel-install/pkg/preinstaller/preinstaller_test.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+			dirSuffix:                "crds",
+			installationResourcePath: resourcePath,
+		}
+
+		// when
+		output, err := i.install(input)
+
+		// then
+		assert.NoError(t, err)
+		assert.True(t, len(output.installed) == 2)
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/preinstaller_test.go:41]: From my understaning we're not aborting installation on first error, but we continue with remaining resources.
It would be nice to see a test case, where we have: "installed: 3, unInstalled: 1". I don't see such test case?

Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/preinstaller_test.go:39]: Great point! Added.

File: parallel-install/pkg/preinstaller/resourcemanager_test.go (hunk 10 above)
Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/resourcemanager_test.go]: I have introduced an exemplary resource.

File: parallel-install/pkg/preinstaller/resourcemanager_test.go (hunk 12 above)
Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/resourcemanager_test.go]: I have introduced an exemplary resource.

File: parallel-install/pkg/preinstaller/preinstaller.go (hunk 3 above)
Comment (clebs) [reply parallel-install/pkg/preinstaller/preinstaller.go:88]: The parallel installation API does offer kyma deletion: https://github.com/kyma-incubator/hydroform/blob/6d2f2ab07d912e5a880c858a502ef17e8fddc21d/parallel-install/pkg/deployment/deletion.go#L26

Therefore I guess removing the CRDs needs to be supported.

File: parallel-install/pkg/preinstaller/resourcemanager.go [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+	return nil
+}
+
+func (c *DefaultResourceManager) GetResource(resourceName string, resourceSchema schema.GroupVersionResource) (obj *unstructured.Unstructured, err error) {
+	obj, err = c.getResource(resourceName, resourceSchema)
+	if err != nil {
+		notFoundError := "not found"
+		matched, _ := regexp.MatchString(notFoundError, err.Error())
+		if matched {
+			c.log.Infof("Resource %s was not found. Skipping additional retries of getting resource.", resourceName)
```
Comment (Tomasz-Smelcerz-SAP) [inline parallel-install/pkg/preinstaller/resourcemanager.go:65]:
```suggestion
			c.log.Infof("Resource %s was not found.", resourceName)
```

**Review by Tomasz-Smelcerz-SAP — APPROVED** [review]

File: parallel-install/pkg/preinstaller/preinstaller.go (hunk 3 above)
Comment (mjakobczyk) [reply parallel-install/pkg/preinstaller/preinstaller.go:88]: @clebs I have created a follow-up issue for this case - https://github.com/kyma-incubator/hydroform/issues/214

**Review by klaudiagrz — APPROVED** [review]

File: parallel-install/pkg/preinstaller/preinstaller.go (hunk 3 above)
Comment (clebs) [reply parallel-install/pkg/preinstaller/preinstaller.go:88]: I am not sure this is fine @mjakobczyk. This PR just broke the current contract of the parallel installation API.
This time it is more or less "OK" t do this because this library is still experimental.

Nevertheless, this will be running everywhere very soon and making breaking changes like this will not be an option.
Please always consider what the current API offers and how changes affect consumers.
